[Services map] Update redirection/Focus field rework

TackAdam opened the pull request:
### Description
Updated the focus field to no longer apply a filter, it now just renders the connected services and allows the user to investigate using the focus window. The drop down now reflects the connected viewable services to switch to. The reset icon now clears the existing filter if it matches the focus field and is currently applied. The focus field is now disabled while a filter of an existing service is active.


https://github.com/user-attachments/assets/e14b2a20-f04c-4d14-95fa-9a3924392717



Clicking a service currently causes the title clicked to be set as a filter.

https://github.com/user-attachments/assets/0acd9974-5ac6-4dc6-a5e7-976dce758732


Updated behavior now re-directs the user to the corresponding service view page

https://github.com/user-attachments/assets/cd27d3ca-88c6-4866-bf3f-4dcc4a5e0524

### Issues Resolved

### Check List
- [ ] New functionality includes testing.
  - [ ] All tests pass, including unit test, integration test and doctest
- [ ] New functionality has been documented.
  - [ ] New functionality has javadoc added
  - [ ] New functionality has user manual doc added
- [ ] Commits are signed per the DCO using --signoff

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

ps48 commented on public/components/trace_analytics/components/services/services_table.tsx:
```diff
@@ -65,20 +68,15 @@ export function ServicesTable(props: ServicesTableProps) {
-      inverted: false,
-      disabled: false,
-    });
+    window.location.href = generateServiceUrl(serviceName, dataSourceMDSId[0].id);
```
The change looks good to me. Can you please add cypress tests and / or unit tests for this? Thanks!

TackAdam replied:
Added jest test 
it('redirects to the correct URL when the service link is clicked',

ps48 commented on public/components/trace_analytics/components/services/service_view.tsx:
```diff
@@ -76,6 +77,18 @@ export function ServiceView(props: ServiceViewProps) {
   const [redirect, setRedirect] = useState(false);
   const [actionsMenuPopover, setActionsMenuPopover] = useState(false);
 
+  let serviceId: string | null = null;
+
+  try {
+    const location = useLocation();
+    const params = new URLSearchParams(location?.search || '');
+    serviceId = params.get('serviceId');
+  } catch (error) {
+    serviceId = null;
+  }
```
Can we use hooks and state here to make it cleaner?

```
  const location = useLocation();

  useEffect(() => {
    try {
      const params = new URLSearchParams(location.search);
      const id = params.get('serviceId');
      setServiceId(id);
    } catch (error) {
      setServiceId(null);
    }
  }, [location]);
  ```

TackAdam replied:
Done. Thanks for the feedback =)

ps48 commented on public/components/trace_analytics/components/services/services_table.tsx:
```diff
@@ -65,20 +68,15 @@ export function ServicesTable(props: ServicesTableProps) {
-      inverted: false,
-      disabled: false,
-    });
+    window.location.href = generateServiceUrl(serviceName, dataSourceMDSId[0].id);
```
Have we test the `generateServiceUrl` with workspaces enabled?

TackAdam replied:
Yes tested with enabled and disabled. Video demo above is in a workspace.

joshuali925 commented on public/components/trace_analytics/components/common/plots/service_map.tsx:
```diff
@@ -81,6 +83,10 @@ export function ServiceMap({
+  filters: FilterType[];
+  setFilters:
+    | React.Dispatch<React.SetStateAction<FilterType[]>>
+    | ((filters: FilterType[]) => void);
```
curious why are both needed?

TackAdam replied:
Thought i needed to cover other conditions, removed it and seems to work fine.
Thanks

joshuali925 commented on public/components/trace_analytics/components/common/plots/service_map.tsx:
```diff
@@ -328,33 +291,36 @@ export function ServiceMap({
+          ticks,
+          service,
+          serviceMap[service]?.relatedServices,
+          true // Enable filtering to focus on connected nodes
```
i didn't look into `getServiceMapGraph`, but seems `service` and `serviceMap[service]?.relatedServices` are used for filtering, and they will be absent if we don't want filtering. Then is this boolean required?

TackAdam replied:
The focus functionality is not applying a filter so for the new experience these are required to change the options that show up using the focus functionality.

Without:

https://github.com/user-attachments/assets/d560e8d1-83c6-4c98-ad40-6b1024dce7f5

With: 

https://github.com/user-attachments/assets/b5fa8fef-eab6-416d-bf5c-b53cfc80607c

joshuali925 commented on public/components/trace_analytics/components/common/plots/service_map.tsx:
```diff
@@ -404,14 +366,16 @@ export function ServiceMap({
+        showRelatedServices
       )
     );
   }, [serviceMap, idSelected]);
```
do we need `focusedService` and `filterByCurrService` in dependency array?

TackAdam replied:
I don't believe so

https://github.com/user-attachments/assets/d97625b4-69e0-4390-a828-a088c6d682da

paulstn replied:
We should include every dependency in the dependency array even if it seems to pointlessly re-render as to avoid 'stale' states that might lead to bugs that would be difficult to fix.
React useEffect doc: https://react.dev/reference/react/useEffect

joshuali925 commented on public/components/trace_analytics/components/services/service_view.tsx:
```diff
@@ -76,6 +77,18 @@ export function ServiceView(props: ServiceViewProps) {
+  let serviceId: string | null = null;
+
+  try {
+    const location = useLocation();
```
> Location is causing the service view page to not render as it is not defined, so it needs to be within the "Try".

do you have more details on what the error is? is it happening when calling `useLocation`?

TackAdam replied:
Resolved with moving the flyout component into router.

joshuali925 commented on public/components/trace_analytics/components/services/service_view.tsx:
```diff
@@ -76,6 +77,18 @@ export function ServiceView(props: ServiceViewProps) {
   const [redirect, setRedirect] = useState(false);
   const [actionsMenuPopover, setActionsMenuPopover] = useState(false);
 
+  let serviceId: string | null = null;
```
should this be a state?

ps48 replied:
➕

TackAdam replied:
Changed to a state.

ps48 commented on public/components/trace_analytics/components/services/__tests__/service_view.test.tsx:
```diff
@@ -7,39 +7,51 @@ import { configure, shallow } from 'enzyme';
+jest.mock('../../../../../../test/__mocks__/coreMocks', () => ({
+  coreStartMock: {
+    chrome: { setBreadcrumbs: jest.fn() },
+    http: { post: jest.fn() },
```
Can we please add mocks with some actual data (both empty and non-empty response)?

TackAdam replied:
Cypress added for service map 
Tests focus functionality in Service map
Testing the re-renders and filtering

Added 
import toJson from 'enzyme-to-json';
To service map test

ps48 commented on public/components/trace_analytics/components/common/plots/service_map.tsx:
```diff
@@ -443,35 +407,57 @@ export function ServiceMap({
+                        if (!isFocusBarDisabled) {
+                          const newValue = e.target.value;
+                          setQuery(newValue);
+                          if (newValue === '') {
+                            if (focusedService) {
+                              setGraphKey((prevKey) => prevKey + 1);
+                              setQuery('');
+                              onFocus(focusedService);
+                            } else {
+                              setGraphKey((prevKey) => prevKey + 1);
+                              setQuery('');
+                              onFocus('');
+                            }
+                          }
```
May be move repeated line outside this `if`:

```suggestion
                          if (newValue === '') {
                           setGraphKey((prevKey) => prevKey + 1);
                           setQuery('');
                           focusedService ?  onFocus(focusedService) : onFocus('');
                          }
```

TackAdam replied:
Changed too     
if (newValue === '') {
                            setGraphKey((prevKey) => prevKey + 1);
                            setQuery('');
                            onFocus(focusedService || '');
                          }

ps48 replied:
Thanks!

ps48 commented on public/components/trace_analytics/components/common/plots/__tests__/service_map.test.tsx:
```diff
@@ -3,25 +3,243 @@
+};
+Object.defineProperty(global, 'crypto', { value: crypto });
+
+const mockContext = ({
+  canvas: document.createElement('canvas'),
+  fillRect: jest.fn(),
+  clearRect: jest.fn(),
```
bigger mocks can be moved to `test/contants`

TackAdam replied:
Moved to MOCK_CANVAS_CONTEXT in constants file

TackAdam commented:
Updated applications to no longer filter on name click. Instead it opens the fly-out for details. Also removed the actions column when in the applications page.

https://github.com/user-attachments/assets/d6ac64f5-a936-4119-9d9f-baed40821e1f

ps48 commented:
> Updated applications to no longer filter on name click. Instead it opens the fly-out for details. Also removed the actions column when in the applications page.
> Screen.Recording.2024-12-06.at.11.25.15.AM.mov

Does the Applications trace navigation work as expected?

TackAdam commented:
> > Updated applications to no longer filter on name click. Instead it opens the fly-out for details. Also removed the actions column when in the applications page.
> > Screen.Recording.2024-12-06.at.11.25.15.AM.mov
> 
> Does the Applications trace navigation work as expected?

Clicking Trace applies the filter and switches to the Traces and Spans  tab.



https://github.com/user-attachments/assets/bd1e19f5-5995-4f6f-a2c1-fa82b72a935e

ps48 commented on public/components/trace_analytics/components/common/plots/__tests__/service_map.test.tsx:
```diff
@@ -98,9 +99,18 @@ describe('ServiceMap Component', () => {
     const wrapper = mount(<ServiceMap {...defaultProps} />);
-    expect(wrapper).toMatchSnapshot();
+    wrapper.update();
+
```
Should we wait for the canvas to appear, so that map renders in the snapshot?

ps48 commented:
Also, looks like this test `1) "before each" hook for "Verifies tree view expand arrow functionality"` is flaky we should solve this immediately in a follow up PR.

ps48 commented:
I see one failure in test because the nodes changes position. We do have a set seed which makes sure the graph has nodes on the position. Need to check what's causing the change cc: @TackAdam

TackAdam commented:
> I see one failure in test because the nodes changes position. We do have a set seed which makes sure the graph has nodes on the position. Need to check what's causing the change cc: @TackAdam

The position is consistent but it seems cypress environment used here might be causing the physics to behave slightly different than we see testing locally on original layout. As the point of this test is to ensure the canvas is rendering the nodes my solution is to first focus on a node to cause less to be rendered and ensure that it matches the layout we see locally. This still ensures the nodes are displayed and are intractable. Ran multiple times and seems to be working with my fix.

ps48 commented on .cypress/integration/trace_analytics_test/trace_analytics_services.spec.js:
```diff
@@ -149,7 +149,7 @@ describe('Testing service view', () => {
   });
 });
-
+//adam
```
Should remove this.

ps48 commented on .cypress/integration/trace_analytics_test/trace_analytics_services.spec.js:
```diff
@@ -171,7 +172,75 @@ describe('Testing Service map', () => {
+    cy.get('.vis-network canvas').should('have.attr', 'height').and('not.eq', '0');
+  });
+
+  it.only('Click on a node to see the details', () => {
```
Should remove `only` here, this is causing only 1 test to run: https://github.com/opensearch-project/dashboards-observability/actions/runs/12216555045/job/34080417940#step:26:118

ps48 commented on .cypress/integration/trace_analytics_test/trace_analytics_services.spec.js:
```diff
@@ -171,7 +172,75 @@ describe('Testing Service map', () => {
+    cy.get('.euiText.euiText--medium .panel-title').contains('Service map');
+    cy.get('.vis-network canvas').should('exist');
+
+    // Focus on "order" by selecting the first option
+    cy.get('.euiFormLabel.euiFormControlLayout__prepend').contains('Focus on').should('exist');
+    cy.get('[placeholder="Service name"]').click();
+    cy.get('.euiSelectableList__list li').eq(0).click();
+
+    // ensure rendering is complete before node click, replace wait
+    cy.get('[data-text="Errors"]').click();
+    cy.contains('60%');
+    cy.get('[data-text="Duration"]').click();
+    cy.contains('100');
```
I have two question on this: 

1) We usually follow one test per it. Looks like we are mixing the focus test and the node click to view details test. This helps us to pin point the issue in the code quickly. Should we keep them independent?

2) Without focus, Why is the click on node with the `x, y` coordinates not working? Is there a possibility in our current code that the graph rendering can change shape on page refresh?

TackAdam replied:
1. I don't think this can be avoided for this particular test as we need the behavior to work locally and on this test unit. If the canvas fails to render the nodes, having a focus will also have the same error so this test still serves its purpose. If focus was broken then this would also fail but the error would still be identifiable.

2. Without focus it is failing because the environment cypress is using is applying the seed differently. (Below screenshot showing its layout vrs the local). The click fails because there is no node at the location in the layout that the cypress version non local is running. The service map will always be the same for the user regardless of refresh. My change works as cypress and local have the same layout.

Local:
<img width="995" alt="LocalTest" src="https://github.com/user-attachments/assets/a20f489c-021c-4c57-8d73-31a0ae6f1e12">

Cypress (Taken from recording of the failure):
<img width="1274" alt="CypressTest" src="https://github.com/user-attachments/assets/64d0e238-7669-4153-9715-643c27155691">

Local and Cypress with Focus
<img width="902" alt="Screenshot 2024-12-07 at 4 21 29 PM" src="https://github.com/user-attachments/assets/bcfb1485-bafc-431a-95fe-d6613ec682d4">

ps48 replied:
> Without focus it is failing because the environment cypress is using is applying the seed differently. 

Are we assuming that the cypress issue won't be replicated on user's end?  We need to understand why is cypress seeing a different graph.
